### ISSUE TITLE
Evaluate other config storage options

### ISSUE THREAD

**Issue by joeholley** [GoogleCloudPlatform/open-match#42]
As mentioned in the docs, right now the 'common' config used by various components is in a json file that gets added to the component containers at build time.  This was always intended as an interim solution, we should discuss and evaluate moving config to a more 'kubernetes-native' format.
## Requirements
- Needs to be accessible by all components
- Should be able to be set using 'standard' resource/file formats like json or yaml

## Nice to have 
- Would be good to have some mechanism for securing values in the config.  For example, how k8s configmaps can include values from k8s secrets.

## Unneeded
- Update at runtime.  Expectation is that the items stored in this config are the kind you read at startup and don't update.  
- Ability to store multiple 'configs' or config 'versioning' within OM.  Expectation is that you will keep your configs alongside your code in your repo, and 'apply' them when you deploy an OM instance.

## Some eval criteria
In the past we've discussed kubernetes ConfigMaps, and those might be a good fit but someone needs to evaluate them.  This isn't an exhaustive list, just what springs to mind immediately.
- How easy are they to manage? Do they require a new set of operator knowledge that we don't expect OM users to have yet?
- What is the mechanism for the components 'reading' the config if it's in a configmap?  
- Are components talking directly to the k8s API to get it, or is it getting sent via setting env vars in the component's k8s resource definitions?  
  - If talking directly to the API, are we putting a requirement on customizable components that they include the necessary components to read from the k8s API (modules in their language of choice, etc).  
  - How does this impact running components outside of k8s (i.e. dev env on local machines?)
- What happens if the config isn't there yet at startup?

**Comment by ihrankouski** [comment]
It's a nice idea at first look:
- We'll have a single source of config in OM cluster and we'll be able to mount it at a any path into containers as a file.
- The container mount will receive auto updates when ConfigMap data is changed. There's a way to prevent it: using `subPath` feature.
- Also it's possible to split configuration into separate smaller specific files but have them deployed in the same ConfigMap still. And then mount specific parts of the ConfigMap at specific paths into conatiners.

Some questions arise however:

1. Feels like we'd better keep Redis username & password in a Secret. But at the moment that's not possible to inject the data of a Secret into the data of ConfigMap using only default k8s tools. So if we decide to store Redis credentials as a Secret we'll have to pass it to containers as env vars.

2. Other thing is the way how ConfigMaps are created and updated in k8s:
	- Start like this: 
	  `kubectl create configmap om-configmap --from-file=config/matchmaker_config.json`
	- Then every time you update the config file locally you'll have to do something like this to deploy the fresh version:
	  `kubectl create configmap om-configmap --from-file=config/matchmaker_config.json --dry-run -o yaml | kubectl replace configmap om-configmap -f -`

Alternatively it's possible to do `kubectl apply -f configmap.yaml` both for creation and update but to achieve this we need to keep the config data inside the ConfigMap resource definition:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: om-configmap
data:
  matchmaker_config.json: |
    {
      "debug": false,
      ...
    }
```

Some tools like *kustomize* offer [ways to keep config data separate from ConfigMap definition](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/kustomization.yaml#L69) and still be able to `kubectl apply -f ...`.
But then we end up managing config files for these tools... 
Well, maybe kustomize will also help with substitution of credentials into configmap (not sure for now) though the credentials will become "declassified" then...

Answering some other your questions:

> How does this impact running components outside of k8s (i.e. dev env on local machines?

Config maps work fine on minikube. 

> What happens if the config isn't there yet at startup?

The pods referring it won't start.

**Comment by joeholley** [comment]
Yeah, I think that we'll deal with the `kubectl replace` command and just document it heavily.  We should break out Redis credentials to a secret as mentioned.  We'll also want to take this opportunity to move the config file from `json` to `yaml` so we can start adding comments; there are a lot of options in there and we need to start having them documented in the file.  Part of that will be updating the `config/config.go:ReadConfig()` function to look for `yaml` as well.

**Comment by joeholley** [comment]
Will be included in 0.4.0 targeted for release in late Feb 2019

**Comment by jeremyje** [comment]
The configs were removed from the docker images, there's a now a ConfigMap created that will mount config/matchmaker_config.yaml.

Changes that resolved this:
https://github.com/GoogleCloudPlatform/open-match/pull/104
https://github.com/GoogleCloudPlatform/open-match/pull/81
